app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -113,12 +113,28 @@ def save_child_inventory(obj, hashes, child_keys, *args)
 
   def store_ids_for_new_records(records, hashes, keys)
     keys = Array(keys)
-    hashes.each do |h|
-      r = records.detect { |r| keys.all? { |k| r.send(k) == r.class.type_for_attribute(k.to_s).cast(h[k]) } }
-      h[:id] = r.id
+    # Lets first index the hashes based on keys, so we can do O(1) lookups
+    record_index = {}
+    records.uniq.each do |record|
+      record_index[build_index_from_record(keys, record)] = record
+    end
+
+    record_class = records.first.class
+
+    hashes.each do |hash|
+      record = record_index[build_index_from_hash(keys, hash, record_class)]
+      hash[:id] = record.try(:id)
```

Comment:
ah, seems like the association.push is not saving the record and the uniq is somehow filtering them out

so it could be this is actually able to fill the refresh only in the second pass?